### DISCUSSION TITLE
Custom ENV files fed in via dotenv?

### DISCUSSION THREAD

**Discussion by jsheffers** [cawa-93/vite-electron-builder#463] [Q&A] [answered]
I'm migrating my project from electron-webpack and I'm running into an issue with injecting certain ENV values into this boilerplate. Here is a little info on how we're currently doing it. We have ENV files that we pass into our electron instance via yarn script commands, so it might look like the following.

`yarn start:app1:dev` or `yarn start:app1:app`

This runs the following scripts defined in package.json. in order to select the correct ENV file.

```
"scripts": {
 "env:app1:dev": "dotenv -e ./config/.env.app1.dev",
 "start:app1:dev": "yarn env:app1:dev electron-webpack dev",
}
```

Then in the `main/index.js` I'd use the `import.meta.env.VITE_URL` to load the proper URL in the Electron window.

I tried using the same concept, but it doesn't seem like Vite allows this.  Does anyone have any insight into this? As an FYI, I also removed Typescript from the template with the instructions from the maintainer

**Comment by cawa-93** [comment, marked as answer]
First, make sure that your environment variables are available when processing the configuration. Try adding something like follow to `packages/main/vite.config.js`

```js
console.log(proccess.env.VITE_URL)
```

Previously, Vite was able to read environment variables simply from `proccess.env` in runtime. But I'm not sure if this behavior has changed.

**Reply by jsheffers** [reply]
The only VITE variable I see in process.env is `VITE_DEV_SERVER_URL: 'http://localhost:3000/'` which I assume is set in watch.js:

```js
process.env.VITE_DEV_SERVER_URL = `${protocol}//${host}:${port}${path}`;
```

However, if I put the vars in `.env.development` I am able to access them within the renderer. Can I create arbitrary modes somehow without Typescript?

**Reply by jsheffers** [reply]
I guess it boils down to the question of whether Vite env files are only based on the MODE or NODE_ENV.

**Reply by cawa-93** [reply]
This may mean that your setup is not loading variables into the environment, or they may not have the correct prefix. By default, vite processes only variables that begin with `VITE_`. The TS does not play any role here. 

I think you could spread all your `.env` files to individual directories and set the appropriate directory in the configuration

https://github.com/cawa-93/vite-electron-builder/blob/fb6aab1303181a165de3bbd9cdd099c50887ab4b/packages/main/vite.config.js#L15

https://vitejs.dev/config/#envdir

Although this is a weird workaround

**Reply by cawa-93** [reply]
> 
> 
> I guess it boils down to the question of whether Vite env files are only based on the MODE or NODE_ENV.

Check this 

https://github.com/vitejs/vite/blob/5cf4e69cd3afc7f960e02072171c7c441747e8f0/packages/vite/src/node/config.ts#L338-L343

https://github.com/vitejs/vite/blob/5cf4e69cd3afc7f960e02072171c7c441747e8f0/packages/vite/src/node/config.ts#L974-L1034

**Reply by jsheffers** [reply]
Thanks for the follow-up @cawa-93!  I ended up going this route:

```
"scripts": {
    "env:app1:dev": "MODE=app1.dev",
    "start:app1:dev": "yarn env:app1:dev node scripts/watch.js",
    "build:app1:dev": "electron-builder --mac --win --config ./config/app1.dev.yaml",
    "env:app1:app": "MODE=app1.app",
    "start:app1:app": "yarn env:app1:app node scripts/watch.js",
    "build:app1:app": "electron-builder --mac --win --config ./config/app1.app.yaml",
}
```

I know this is an odd use-case, but we use one Electron instance to build out multiple apps (pointed at different URL's) as well as the renderer process for some things.

These scripts in combination with env files named `.env.app1.dev`, `.env.app1.app` fix my issue.
